crates/gpui/src/platform/windows/destination_list.rs

```diff
@@ -0,0 +1,189 @@
+            )),
+        }
+    }
+}
+
+// https://github.com/microsoft/Windows-classic-samples/blob/main/Samples/Win7Samples/winui/shell/appshellintegration/RecipePropertyHandler/RecipePropertyHandler.cpp
+pub(crate) fn update_jump_list(
+    entries: &[&Vec<String>],
+    dock_menus: &[DockMenuItem],
+) -> anyhow::Result<Option<Vec<Vec<String>>>> {
+    let (list, removed) = create_destination_list()?;
+    add_recent_folders(&list, entries, removed.as_ref())?;
+    add_dock_menu(&list, dock_menus)?;
+    unsafe { list.CommitList() }?;
+    Ok(removed)
+}
+
+const PKEY_TITLE: PROPERTYKEY = PROPERTYKEY {
+    fmtid: GUID::from_u128(0xf29f85e0_4ff9_1068_ab91_08002b27b3d9),
+    pid: 2,
+};
+
+const PKEY_LINK_ARGS: PROPERTYKEY = PROPERTYKEY {
+    fmtid: GUID::from_u128(0x436f2667_14e2_4feb_b30a_146c53b5b674),
+    pid: 100,
+};
+
+fn create_destination_list() -> anyhow::Result<(ICustomDestinationList, Option<Vec<Vec<String>>>)> {
+    let list: ICustomDestinationList =
+        unsafe { CoCreateInstance(&DestinationList, None, CLSCTX_INPROC_SERVER) }?;
+
+    let mut slots = 0;
+    let user_removed: IObjectArray = unsafe { list.BeginList(&mut slots) }?;
+
+    let count = unsafe { user_removed.GetCount() }?;
+    if count == 0 {
+        return Ok((list, None));
+    }
+
+    let mut removed = Vec::with_capacity(count as usize);
+    for i in 0..count {
+        let shell_link: IShellLinkW = unsafe { user_removed.GetAt(i)? };
+        let store: IPropertyStore = shell_link.cast()?;
+        let argument = unsafe { store.GetValue(&PKEY_LINK_ARGS)? };
+        let args = argument
+            .to_string()
+            .split_whitespace()
+            .map(|s| s.trim_matches('"').to_string())
+            .collect_vec();
+
+        removed.push(args);
+    }
+
+    Ok((list, Some(removed)))
+}
+
+fn add_dock_menu(list: &ICustomDestinationList, dock_menus: &[DockMenuItem]) -> anyhow::Result<()> {
+    unsafe {
+        let tasks: IObjectCollection =
+            CoCreateInstance(&EnumerableObjectCollection, None, CLSCTX_INPROC_SERVER)?;
+        for (idx, dock_menu) in dock_menus.iter().enumerate() {
+            let argument = HSTRING::from(format!("--dock-action {}", idx));
+            let description = HSTRING::from(dock_menu.description.as_str());
+            let display = dock_menu.name.as_str();
+            let task = create_shell_link(argument, description, None, display)?;
+            tasks.AddObject(&task)?;
+        }
+        list.AddUserTasks(&tasks)?;
+        Ok(())
+    }
+}
+
+fn add_recent_folders(
+    list: &ICustomDestinationList,
+    entries: &[&Vec<String>],
+    removed: Option<&Vec<Vec<String>>>,
+) -> anyhow::Result<()> {
+    unsafe {
+        let tasks: IObjectCollection =
+            CoCreateInstance(&EnumerableObjectCollection, None, CLSCTX_INPROC_SERVER)?;
+
+        for folder_path in entries
+            .iter()
+            .filter(|path| !is_item_in_array(path, removed))
```

Comment:
It looks like the interior `&Vec<String>` is being treated like a path? Is there any reason to not use a `&Path` here?